dra-evolution/testdata/pools-two-nodes-dgxa100.yaml

```diff
@@ -0,0 +1,1121 @@
+          name: jpeg-engines
+        - capacity: "1"
+          name: ofa-engines
+        - capacity: 4864Mi
+          name: memory
+        - capacity: "1"
+          name: memory-slices-6
+      - attributes:
+        - name: mig-profile
+          string: 1g.10gb
+        name: mig-1g.10gb-0
+        sharedCapacityConsumed:
+        - capacity: "14"
+          name: multiprocessors
+        - capacity: "1"
+          name: copy-engines
+        - capacity: "1"
+          name: decoders
+        - capacity: 9856Mi
+          name: memory
+        - capacity: "1"
+          name: memory-slices-0
+        - capacity: "1"
+          name: memory-slices-1
+      - attributes:
+        - name: mig-profile
+          string: 1g.10gb
+        name: mig-1g.10gb-2
+        sharedCapacityConsumed:
+        - capacity: "14"
+          name: multiprocessors
+        - capacity: "1"
+          name: copy-engines
+        - capacity: "1"
+          name: decoders
+        - capacity: 9856Mi
+          name: memory
+        - capacity: "1"
+          name: memory-slices-2
+        - capacity: "1"
+          name: memory-slices-3
+      - attributes:
+        - name: mig-profile
+          string: 1g.10gb
+        name: mig-1g.10gb-4
+        sharedCapacityConsumed:
+        - capacity: "14"
+          name: multiprocessors
+        - capacity: "1"
+          name: copy-engines
+        - capacity: "1"
+          name: decoders
+        - capacity: 9856Mi
+          name: memory
+        - capacity: "1"
+          name: memory-slices-4
+        - capacity: "1"
+          name: memory-slices-5
+      - attributes:
+        - name: mig-profile
+          string: 1g.10gb
+        name: mig-1g.10gb-6
+        sharedCapacityConsumed:
+        - capacity: "14"
+          name: multiprocessors
+        - capacity: "1"
+          name: copy-engines
+        - capacity: "1"
+          name: decoders
+        - capacity: 9856Mi
+          name: memory
+        - capacity: "1"
+          name: memory-slices-6
+        - capacity: "1"
+          name: memory-slices-7
+      sharedCapacity:
+      - capacity: 40Gi
+        name: memory
+      - capacity: "98"
+        name: multiprocessors
+      - capacity: "7"
+        name: copy-engines
+      - capacity: "5"
+        name: decoders
+      - capacity: "1"
+        name: jpeg-engines
+      - capacity: "1"
+        name: ofa-engines
+      - capacity: "1"
+        name: memory-slices-0
+      - capacity: "1"
+        name: memory-slices-1
+      - capacity: "1"
+        name: memory-slices-2
+      - capacity: "1"
+        name: memory-slices-3
+      - capacity: "1"
+        name: memory-slices-4
+      - capacity: "1"
+        name: memory-slices-5
+      - capacity: "1"
+        name: memory-slices-6
+      - capacity: "1"
+        name: memory-slices-7
+    devices:
+    - attributes:
+      - name: uuid
+        string: GPU-e859d6ce-7f9c-4f46-b122-f3df7ebf9f06
+      - bool: true
+        name: mig-capable
```

Comment:
Does this get inherited by the partitions in the shape?  IOW, do the MIG partitions indicate "mig-capable"?

Reply:
Yeah. That's an issue. Similar to the issue Kevin raised. We might need some attribute "inheritance" decorations which I don't love.

Reply:
This is what I meant "class" vs. "instance".  I'll try to sketch out what I mean.

Reply:
Here, I hacked up just a couple examples.  Doing this makes me ask - doesn't each partition of each device need a unique name, which is used from claim status?

```diff
--- /home/thockin/Downloads/orig.yaml	2024-06-15 14:26:21.270853559 -0700
+++ /home/thockin/Downloads/th-example.yaml	2024-06-15 14:44:01.254635103 -0700
@@ -495,19 +495,77 @@
     - attributes:
       - name: uuid
         string: GPU-e859d6ce-7f9c-4f46-b122-f3df7ebf9f06
-      - bool: true
-        name: mig-capable
-      - bool: false
-        name: mig-enabled
       name: gpu-0
+      partitions: ## NOTE: instantiates the partitions defined in the shape
+      - whole:
+        attributes:
+        - bool: true
+          name: mig-capable
+      ## NOTE: maybe just elide these if they are actually empty, or maybe
+      ## they all will actually have instancy things.  Do they each need a
+      ## name, which is referenced from claim status?
+      - mig-1g.5gb-0: {}
+      - mig-1g.5gb-1: {}
+      - mig-1g.5gb-2: {}
+      - mig-1g.5gb-3: {}
+      - mig-1g.5gb-4: {}
+      - mig-1g.5gb-5: {}
+      - mig-1g.5gb-6: {}
+      - mig-2g.10gb-0: {}
+      - mig-2g.10gb-2: {}
+      - mig-2g.10gb-4: {}
+      - mig-3g.20gb-0: {}
+      - mig-3g.20gb-4: {}
+      - mig-4g.20gb-0: {}
+      - mig-7g.40gb-0: {}
+      - mig-1g.5gb-me-0: {}
+      - mig-1g.5gb-me-1: {}
+      - mig-1g.5gb-me-2: {}
+      - mig-1g.5gb-me-3: {}
+      - mig-1g.5gb-me-4: {}
+      - mig-1g.5gb-me-5: {}
+      - mig-1g.5gb-me-6: {}
+      - mig-1g.10gb-0: {}
+      - mig-1g.10gb-2: {}
+      - mig-1g.10gb-4: {}
+      - mig-1g.10gb-6: {}
     - attributes:
       - name: uuid
         string: GPU-f87f3708-c8cf-4418-933b-2cb12f1f7d7c
-      - bool: true
-        name: mig-capable
-      - bool: false
-        name: mig-enabled
       name: gpu-1
+      partitions: ## NOTE: instantiates the partitions defined in the shape
+      - whole:
+        attributes:
+        - bool: true
+          name: mig-capable
+      ## NOTE: maybe just elide these if they are actually empty, or maybe
+      ## they all will actually have instancy things.  Do they each need a
+      ## name, which is referenced from claim status?
+      - mig-1g.5gb-0: {}
+      - mig-1g.5gb-1: {}
+      - mig-1g.5gb-2: {}
+      - mig-1g.5gb-3: {}
+      - mig-1g.5gb-4: {}
+      - mig-1g.5gb-5: {}
+      - mig-1g.5gb-6: {}
+      - mig-2g.10gb-0: {}
+      - mig-2g.10gb-2: {}
+      - mig-2g.10gb-4: {}
+      - mig-3g.20gb-0: {}
+      - mig-3g.20gb-4: {}
+      - mig-4g.20gb-0: {}
+      - mig-7g.40gb-0: {}
+      - mig-1g.5gb-me-0: {}
+      - mig-1g.5gb-me-1: {}
+      - mig-1g.5gb-me-2: {}
+      - mig-1g.5gb-me-3: {}
+      - mig-1g.5gb-me-4: {}
+      - mig-1g.5gb-me-5: {}
+      - mig-1g.5gb-me-6: {}
+      - mig-1g.10gb-0: {}
+      - mig-1g.10gb-2: {}
+      - mig-1g.10gb-4: {}
+      - mig-1g.10gb-6: {}
     - attributes:
       - name: uuid
         string: GPU-d197c034-8516-4dda-a2cd-67a46457a454
```

Reply:
Yes, but they don't need to be stored. They are just the concatenation of the device name with the partition name. 

If there are N partitions in shape, and M devices in the slice, the scheduler will be presented with N * M "devices" for satisfying claims from the slice. 

This API is just what we need to store.

Reply:
Unless we need to set values on a per device, per partition basis, these instances are redundant.

Reply:
> Yes, but they don't need to be stored. They are just the concatenation of the device name with the partition name.

As long as we are clear in spec, I am OK with that.  I didn't see it written.  Does that mean "gpu-0/whole" is the name?  Or should the "whole" partition be specially named as ""?  

> Unless we need to set values on a per device, per partition basis, these instances are redundant

But we DO need to.  At least one of them: "mig-capable" applies to the "whole" partition and none of the others.  I bet there are more.  I think I would be fine to elide empty instances here, and treat this like a "last level override" for cases where the instance needs instance attribs.

Reply:
> As long as we are clear in spec, I am OK with that. I didn't see it written. Does that mean "gpu-0/whole" is the name? Or should the "whole" partition be specially named as ""?

See  https://github.com/kubernetes-sigs/wg-device-management/pull/31/files#diff-a53461b1c5a20f7a5b62cecae4b45445b5c874d8d192093ade65eb9cc06e3ec1R112

It would be `gpu-0-whole`. We could do `""` maybe. The weird thing is that it is still a "partition". I wonder if we can just make "whole device is allocatable" a flag and just use the device name? A sort of "pseudo" partition? Not sure.

> But we DO need to. At least one of them: "mig-capable" applies to the "whole" partition and none of the others. I bet there are more. I think I would be fine to elide empty instances here, and treat this like a "last level override" for cases where the instance needs instance attribs.

Yes, that could work. We could also include the "whole device" as something special (as alluded to above). Or do both of those things, they are not mutually exclusive.

Reply:
I've pushed a rough outline of my thoughts on (yet another) way to do all of this and will start to refine it in the morning. I wanted to get it out before I go to bed though, in case any initial comments come in overnight that might influence how I push forward on it: https://github.com/kubernetes-sigs/wg-device-management/pull/35

Reply:
I'll try to take a look later tonight